File: BeatLooper/View Controllers/HomeViewController.m
//
//  ViewController.m
//  BeatLooper
//
//  Created by Isaak Meier on 4/2/21.
//

#import "HomeViewController.h"

@interface HomeViewController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *songTableView;
@property (weak, nonatomic) IBOutlet UIButton *editButton;
@property (strong, nonatomic) NSMutableArray *songs;
@property (weak, nonatomic) IBOutlet UIImageView *rainbowMusicBanner;
@property BOOL isAddSongsMode;
@property BOOL rowSelected;
@property NSString *currentlyPlayingSongTitle;

@end

@implementation HomeViewController

- (id)initWithCoordinator:(BLPCoordinator *)coordinator inAddSongsMode:(BOOL)isAddSongsMode {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    if ((self = [storyboard instantiateViewControllerWithIdentifier:@"HomeViewController"])) {
        _coordinator = coordinator;
        _model = [[BLPBeatModel alloc] init];
        _isAddSongsMode = isAddSongsMode;
    }
    return self;
}

- (void)refreshSongsAndReloadData:(BOOL)shouldReloadData {
    NSArray *brandNewSongs = [[self model] getAllSongs];
    [self setSongs:[NSMutableArray arrayWithArray:brandNewSongs]];
    if (shouldReloadData) {
        [[self songTableView] reloadData];
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    [self songTableView].dataSource = self;
    [self songTableView].delegate = self;
    [[self songTableView] registerClass:[UITableViewCell class] forCellReuseIdentifier:@"SongCell"];

    if (self.isAddSongsMode) {
        [self.editButton setHidden:YES];
        [self.rainbowMusicBanner setHidden:YES];
    }
    [self refreshSongsAndReloadData:YES];
}

- (IBAction)editButtonTapped:(id)sender {
    if (self.songTableView.isEditing) {
        [self.songTableView setEditing:NO];
        [self.editButton setTitle:@"Edit" forState:UIControlStateNormal];
    } else {
        [self.songTableView setEditing:YES];
        [self.editButton setTitle:@"Done" forState:UIControlStateNormal];
    }
}

// MARK: UITableView Datasource
- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView
                 cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [self.songTableView dequeueReusableCellWithIdentifier:@"SongCell"];
    Beat *beat = self.songs[indexPath.row];
    cell.textLabel.text = beat.title;
    if ([self.currentlyPlayingSongTitle isEqualToString:beat.title]) {
        UILabel *nowPlayingView = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 120, cell.frame.size.height)];
        [nowPlayingView setText:@"Now Playing"];
        [nowPlayingView setTextColor:UIColor.grayColor];
        UIView *labelHolder = [[UIView alloc] initWithFrame:nowPlayingView.frame];
        labelHolder.backgroundColor = UIColor.clearColor;
        [labelHolder addSubview:nowPlayingView];
        cell.accessoryView = labelHolder;
    } else {
        cell.accessoryView = nil;
    }
    return cell;
    
}


// MARK: UITableView Delegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (self.isAddSongsMode) {
        [self.coordinator addSongToQueue:self.songs[indexPath.row]];
        return;
    }
    Beat *beat = self.songs[indexPath.row];
    if ([self.currentlyPlayingSongTitle isEqualToString:beat.title]) {
        [self.coordinator openPlayerWithoutSong];
    } else {
        // this range encompasses the song we just selected and every song after it.
        NSRange queueRange = NSMakeRange(indexPath.row, self.songs.count - indexPath.row);
        NSIndexSet *indexes = [[NSIndexSet alloc] initWithIndexesInRange:queueRange];
        NSArray *songsForQueue = [NSArray arrayWithArray:[self.songs objectsAtIndexes:indexes]];
        [[self coordinator] openPlayerWithSongs:songsForQueue];
        self.rowSelected = YES;
    }
    [self.songTableView deselectRowAtIndexPath:indexPath animated:NO];
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        Beat *selectedBeat = self.songs[indexPath.row];
        [self.model deleteSong:selectedBeat];
        [self refreshSongsAndReloadData:NO];
        NSArray *indexPaths = @[indexPath];
        [self.songTableView deleteRowsAtIndexPaths:indexPaths withRowAnimation:UITableViewRowAnimationFade];
    }
}

- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)sourceIndexPath toIndexPath:(NSIndexPath *)destinationIndexPath {
    if (sourceIndexPath.row != destinationIndexPath.row) {
        Beat *beatToMove = self.songs[sourceIndexPath.row];
        [self.songs removeObjectAtIndex:sourceIndexPath.row];
        [self.songs insertObject:beatToMove atIndex:destinationIndexPath.row];
    }
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.songs count];
}


// MARK: BLPPlayer Delegate
- (void)currentItemDidChangeStatus:(AVPlayerItemStatus)status {
    // do nothing
}

- (void)didUpdateCurrentProgressTo:(double)fractionCompleted {
    // do nothing
}

- (void)playerDidChangeSongTitle:(nonnull NSString *)songTitle {
    // update the correct tableviewcell with the title
    self.currentlyPlayingSongTitle = songTitle;
    [self.songTableView reloadData];
}

- (void)playerDidChangeState:(BLPPlayerState)state {
    // do nothing
}

- (void)requestProgressBarUpdate {
    // do nothing
}

- (void)requestTableViewUpdate {
    // do nothing 
}

- (void)selectedIndexesChanged:(NSUInteger)count { 
    // do nothing
}

@end

File: BeatLooper/View Controllers/HomeViewController.swift
//
//  HomeViewController.swift
//  BeatLooper 2
//
//  Created by Isaak Meier on 4/2/21.
//  Migrated to Swift
//

import UIKit

class HomeViewController: UIViewController {
    
    // MARK: - IBOutlets
    @IBOutlet weak var songTableView: UITableView!
    @IBOutlet weak var editButton: UIButton!
    @IBOutlet weak var rainbowMusicBanner: UIImageView!
    
    // MARK: - Properties
    var model: BeatModel!
    weak var coordinator: Coordinator?
    private var songs: [Beat] = []
    private var isAddSongsMode: Bool = false
    private var rowSelected: Bool = false
    private var currentlyPlayingSongTitle: String?
    
    // MARK: - Initialization
    init(coordinator: Coordinator, inAddSongsMode: Bool) {
        super.init(nibName: nil, bundle: nil)
        self.coordinator = coordinator
        self.model = BeatModel()
        self.isAddSongsMode = inAddSongsMode
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    // MARK: - Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        refreshSongsAndReloadData(shouldReloadData: true)
    }
    
    // MARK: - Setup
    private func setupUI() {
        songTableView.dataSource = self
        songTableView.delegate = self
        songTableView.register(UITableViewCell.self, forCellReuseIdentifier: "SongCell")
        
        if isAddSongsMode {
            editButton.isHidden = true
            rainbowMusicBanner.isHidden = true
        }
    }
    
    // MARK: - Public Methods
    func refreshSongsAndReloadData(shouldReloadData: Bool) {
        let brandNewSongs = model.getAllSongs()
        songs = brandNewSongs
        if shouldReloadData {
            songTableView.reloadData()
        }
    }
    
    // MARK: - Actions
    @IBAction func editButtonTapped(_ sender: UIButton) {
        if songTableView.isEditing {
            songTableView.setEditing(false, animated: true)
            editButton.setTitle("Edit", for: .normal)
        } else {
            songTableView.setEditing(true, animated: true)
            editButton.setTitle("Done", for: .normal)
        }
    }
}

// MARK: - UITableViewDataSource
extension HomeViewController: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return songs.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = songTableView.dequeueReusableCell(withIdentifier: "SongCell", for: indexPath)
        let beat = songs[indexPath.row]
        cell.textLabel?.text = beat.title
        
        if currentlyPlayingSongTitle == beat.title {
            let nowPlayingView = UILabel(frame: CGRect(x: 0, y: 0, width: 120, height: cell.frame.size.height))
            nowPlayingView.text = "Now Playing"
            nowPlayingView.textColor = .gray
            
            let labelHolder = UIView(frame: nowPlayingView.frame)
            labelHolder.backgroundColor = .clear
            labelHolder.addSubview(nowPlayingView)
            cell.accessoryView = labelHolder
        } else {
            cell.accessoryView = nil
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        return true
    }
    
    func tableView(_ tableView: UITableView, moveRowAt sourceIndexPath: IndexPath, to destinationIndexPath: IndexPath) {
        if sourceIndexPath.row != destinationIndexPath.row {
            let beatToMove = songs[sourceIndexPath.row]
            songs.remove(at: sourceIndexPath.row)
            songs.insert(beatToMove, at: destinationIndexPath.row)
        }
    }
}

// MARK: - UITableViewDelegate
extension HomeViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if isAddSongsMode {
            coordinator?.addSongToQueue(songs[indexPath.row])
            return
        }
        
        let beat = songs[indexPath.row]
        if currentlyPlayingSongTitle == beat.title {
            coordinator?.openPlayerWithoutSong()
        } else {
            // This range encompasses the song we just selected and every song after it
            let queueRange = NSRange(location: indexPath.row, length: songs.count - indexPath.row)
            let indexes = IndexSet(integersIn: queueRange)
            let songsForQueue = Array(songs[indexes])
            coordinator?.openPlayerWithSongs(songsForQueue)
            rowSelected = true
        }
        
        tableView.deselectRow(at: indexPath, animated: false)
    }
    
    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            let selectedBeat = songs[indexPath.row]
            model.deleteSong(selectedBeat)
            refreshSongsAndReloadData(shouldReloadData: false)
            
            let indexPaths = [indexPath]
            tableView.deleteRows(at: indexPaths, with: .fade)
        }
    }
}

// MARK: - PlayerDelegate
extension HomeViewController: PlayerDelegate {
    
    func currentItemDidChangeStatus(_ status: AVPlayerItem.Status) {
        // Do nothing
    }
    
    func didUpdateCurrentProgressTo(_ fractionCompleted: Double) {
        // Do nothing
    }
    
    func playerDidChangeSongTitle(_ songTitle: String) {
        // Update the correct table view cell with the title
        currentlyPlayingSongTitle = songTitle
        songTableView.reloadData()
    }
    
    func playerDidChangeState(_ state: PlayerState) {
        // Do nothing
    }
    
    func requestProgressBarUpdate() {
        // Do nothing
    }
    
    func requestTableViewUpdate() {
        // Do nothing
    }
    
    func selectedIndexesChanged(_ count: Int) {
        // Do nothing
    }
}
